x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/pivot/PivotConfig.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.dataframe.transforms.pivot;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationBuilder;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
+public class PivotConfig implements Writeable, ToXContentObject {
+
+    private static final String NAME = "data_frame_transform_pivot";
+    private static final ParseField GROUP_BY = new ParseField("group_by");
+    private static final ParseField AGGREGATIONS = new ParseField("aggregations");
+
+    private final List<GroupConfig> groups;
+    private final AggregationConfig aggregationConfig;
+
+    private static final ConstructingObjectParser<PivotConfig, Void> PARSER = createParser(false);
+    private static final ConstructingObjectParser<PivotConfig, Void> LENIENT_PARSER = createParser(true);
+
+    private static ConstructingObjectParser<PivotConfig, Void> createParser(boolean ignoreUnknownFields) {
+        ConstructingObjectParser<PivotConfig, Void> parser = new ConstructingObjectParser<>(NAME, ignoreUnknownFields,
+                args -> {
+                    @SuppressWarnings("unchecked")
+                    List<GroupConfig> groups = (List<GroupConfig>) args[0];
+                    AggregationConfig aggregationConfig = (AggregationConfig) args[1];
+                    return new PivotConfig(groups, aggregationConfig);
+                });
+
+        parser.declareObjectArray(constructorArg(),
+                (p, c) -> (GroupConfig.fromXContent(p, ignoreUnknownFields)), GROUP_BY);
+
+        parser.declareObject(constructorArg(), (p, c) -> AggregationConfig.fromXContent(p), AGGREGATIONS);
+
+        return parser;
+    }
+
+    public PivotConfig(final List<GroupConfig> groups, final AggregationConfig aggregationConfig) {
+        this.groups = Objects.requireNonNull(groups);
```

Comment:
Use `ExceptionsHelper.requireNonNull` to get a nicer error message when these fields are left unspecified by the user